src/databricks/sql/thrift_backend.py

```diff
@@ -279,24 +287,41 @@ def attempt_request(attempt):
             # - non-None method_return -> success, return and be done
             # - non-None retry_delay -> sleep delay before retry
             # - error, error_message always set when available
+
+            error = None
+
+            # If retry_delay is None the request is treated as non-retryable
+            retry_delay = None
             try:
                 logger.debug("Sending request: {}".format(request))
                 response = method(request)
                 logger.debug("Received response: {}".format(response))
                 return response
-            except Exception as error:
+            except socket.timeout as err:
```

Comment:
do we retry on socket.timeout?

if so, how do we ensure idempotency of write requests?

Reply:
I answered this exact question in the pull request description:

> By the same logic we know that requests which fail with socket.timeout are idempotent because no connection was made to the server. No matter how many times it runs it does not change data on the server. Therefore it can be safely added to the allow list

Reply:
The more I've researched the more I thought about this I think there is still a gap, like you identified. 

Errno 110 always means EHOSTUNREACHABLE which is de facto idemptoent. 

But socket.timeout can be raised by any socket operation, not just connect(). So I'll need some further rewriting to specifically check what kind of socket operation failed.

If the socket operation that failed is a connection, then we can retry the request because we know nothing was transmitted to the server. 

But if it was a send / write operation, we can't guarantee a command wasn't sent to the server and therefore shouldn't retry.

Reply:
+1
If figuring out what socket operation was done is hard, we could also restrict retries to GetOperationStatus here, like in the previous pr.

Reply:
That's exactly what I've done @sander-goos. I updated the PR description but tl;dr we no retry under exactly three scenarios:

- HTTP Response 401 from server
- HTTP Response 503 from server
- `OSError` was raised by system during `GetOperationStatus` request